### PR TITLE
Emulator support

### PR DESCRIPTION
**Pull request by saleemrashid**
Emulator support with SDL2 or headless mode, persistent flash and UDP transport.

### CODE REVIEW

**Comment by saleemrashid** [comment]
Rebased. Device tests pass in 36 seconds on my laptop which is about 12 times faster than on a physical device.

**Comment by saleemrashid** [comment]
> Does it make sense to move `firmware/udp.c` to `emulator/udp.c`?

While it might make sensible to move some of it (for use in non-firmware parts), some of it should remain in `firmware/` because I would like code from `emulator/` to not depend on anything from `firmware/` (in this case `firmware/usb.h`)

**Comment by saleemrashid** [comment]
@prusnak I think this is an acceptable solution

File: .travis.yml [hunk 1]
```diff
@@ -19,6 +19,23 @@ env:
...
+  include:
+    - addons:
+        apt:
+          packages:
+            - gcc-multilib
+      env:
+        - EMULATOR=1 HEADLESS=1
+        - DEBUG_LINK=1
+      before_script:
+        - sed -i '/hidapi/d' "vendor/python-trezor/requirements.txt"
```
Comment (prusnak) [inline .travis.yml]: python-trezor contains `--disable-hidapi` option in `setup.py`, not sure how we can utilize this when installing via pip, though

Comment (saleemrashid) [reply .travis.yml]: @prusnak I tried that at least 5 different ways and gave up. :laughing: `pip` will not pass provided options to `./setup.py egg_info` which is what it uses to gather dependencies.

Comment (prusnak) [reply .travis.yml]: Ok, glad to know I am the only one who failed on this. Thanks :-)

Comment (saleemrashid) [reply .travis.yml]: @prusnak There is an issue here but I don't imagine it will be fixed https://github.com/pypa/pip/issues/4383

**Comment by prusnak** [comment]
:heart:

**Pull request merged**